fix: support amazon-ecr-credential-helper & check for bucket existence before emptying the bucket

hariohmprasath opened the pull request:
### Summary

#2323 - Depending on the value of `credsStore` attribute (inside docker config file), the code either performs the `docker login` or skips it.
#2070 - Added code to check for the existence of the bucket using `s3.HeadBucket` API before proceed to empty the contents

### Issue number
Fixes  #2323 & #2070

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

hariohmprasath committed:
fix: Code changes to support docker push using amazon-ecr-credential-helper #2323

Depending on the value of `credsStore` attribute (inside docker config file), the code either performs the `docker login` or skips it.

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +161,38 @@ func imageName(uri, tag string) string {
 }
+
+// GetHelperProviderFromDockerCfg will fetch credential store used by docker from config file
+func (c DockerCommand) GetHelperProviderFromDockerCfg() (string) {
```
Quick question: does this function work for windows? Since we support windows maybe we don't want this function to break Windows users.

efekarakus replied:
nit: 
```suggestion
func (c DockerCommand) GetHelperProviderFromDockerCfg() string {
```

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +161,38 @@ func imageName(uri, tag string) string {
 	return fmt.Sprintf("%s:%s", uri, tag)
 }
+
+// GetHelperProviderFromDockerCfg will fetch credential store used by docker from config file
```
```suggestion
// GetHelperProviderFromDockerCfg tries to fetch credential store used by docker from config file.
```

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +161,38 @@ func imageName(uri, tag string) string {
+	return ""
+}
+
+// parseCredFromDockerConfig will unmarshal string to struct
```
Maybe we don't need this comment for private function since it doesn't add a lot of value.

hariohmprasath replied:
Removed

iamhopaul123 commented on internal/pkg/repository/repository.go:
```diff
@@ -54,13 +55,19 @@ func (r *Repository) BuildAndPush(docker ContainerLoginBuildPusher, args *exec.B
+	// Check for usage of amazon-ecr-credential-helper before performing docker login
+	credStore := docker.GetHelperProviderFromDockerCfg()
+
+	// Perform docker login only if credStore not
```
Is this comment incomplete?

hariohmprasath replied:
Updated comments

iamhopaul123 reviewed (commented):
Thank you so much for making this change! This is awesome! Just had some nits and a quick question.

efekarakus commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +161,38 @@ func imageName(uri, tag string) string {
+	// Look into the default locations
+	pathsToTry := []string{"/.docker/config.json", ".dockercfg"}
+	for _, path := range pathsToTry {
+		content, err := ioutil.ReadFile(fmt.Sprintf("%s/%s",os.Getenv("HOME"), path))
```
Will this result in a double "//" such as "<homepath>//.docker/config.json"?
```suggestion
	pathsToTry := []string{filepath.Join(".docker, config.json"), ".dockercfg"}
	for _, path := range pathsToTry {
		content, err := ioutil.ReadFile(filepath.Join(os.Getenv("HOME"), path))
```

efekarakus commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +161,38 @@ func imageName(uri, tag string) string {
+// GetHelperProviderFromDockerCfg will fetch credential store used by docker from config file
+func (c DockerCommand) GetHelperProviderFromDockerCfg() (string) {
+	// Look into the default locations
+	pathsToTry := []string{"/.docker/config.json", ".dockercfg"}
+	for _, path := range pathsToTry {
+		content, err := ioutil.ReadFile(fmt.Sprintf("%s/%s",os.Getenv("HOME"), path))
```
What do you think of storing the environment variable in the `DockerCommand` type to allow writing unit tests for this new method:
```go
// DockerCommand represents docker commands that can be run.
type DockerCommand struct {
	runner
	// Override in unit tests.
	homePath string // new field that can be substituted in tests
	buf *bytes.Buffer
}

// NewDockerCommand returns a DockerCommand.
func NewDockerCommand() DockerCommand {
	return DockerCommand{
		runner: NewCmd(),
		homePath: os.Getenv("HOME"),
	}
}
```

(for an example of unit tests with fake file systems, we use `afero.NewMemMapFs()`: https://github.com/aws/copilot-cli/blob/c2d143dfe781a865ed823984c88065a92836b4c6/internal/pkg/workspace/workspace_test.go#L123-L126)

hariohmprasath replied:
Makes sense and thanks for the suggestion. I have added additional test cases to unit test this method

efekarakus commented on internal/pkg/repository/repository.go:
```diff
@@ -54,13 +55,19 @@ func (r *Repository) BuildAndPush(docker ContainerLoginBuildPusher, args *exec.B
+	credStore := docker.GetHelperProviderFromDockerCfg()
+
+	// Perform docker login only if credStore not
+	if credStore != "ecr-login"{
```
can we move "ecr-login" to a const?
```go
const (
   credStoreECRLogin = "ecr-login"
)
```

hariohmprasath replied:
Done

efekarakus reviewed (commented):
Awesome, this is so good thank you so much for fixing this issue 😍 !   
Some suggestions for enabling writing unit tests for the new functionality 🙏

hariohmprasath committed:
Update internal/pkg/exec/docker.go

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

hariohmprasath committed:
Update internal/pkg/exec/docker.go

Co-authored-by: Penghao He <penghaoh@amazon.com>

hariohmprasath committed:
Update internal/pkg/exec/docker.go

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

hariohmprasath committed:
Incorporating code review changes

hariohmprasath committed:
Incorporating code review feedback and updates to ecr credential lookup

Based on the documentation available [here](https://github.com/awslabs/amazon-ecr-credential-helper), `ecr-login` can be enabled globally using `credsStore` or on individual registry level. The recent code update considers both of this scenarios before letting it perform `docker login`

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +172,51 @@ func imageName(uri, tag string) string {
+func userHomeDirectory() string {
+	home, err := os.UserHomeDir()
+	if err!= nil{
+		println("authenticate to ECR: %w", err)
```
Can we remove this line? We don't log outside of the `cli` pkg. Or you might want to define a new error and catch that in the `cli` pkg.

hariohmprasath replied:
Removed

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +172,51 @@ func imageName(uri, tag string) string {
+func (c DockerCommand) IsEcrCredentialHelperEnabled(uri string) bool {
+	// Make sure the program is able to obtain the home directory
+	splits := strings.Split(uri, "/")
+	if c.homePath != "" && len(splits) > 0 {
```
Hmm why not just 
```suggestion
	homePath := userHomeDirectory()
	if homePath != "" && len(splits) > 0 {
```
I don't see `homePath` being used in the other places.

hariohmprasath replied:
Its done to enable unit testing, so part of unit test code we can control this path, like below:

```go
s := DockerCommand{
  runner: mockRunner,
  buf:    tc.inBuffer,
  homePath: "test/copilot",
}
```

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +172,51 @@ func imageName(uri, tag string) string {
+	return false
+}
+
+func parseCredFromDockerConfig(config []byte) (*dockerConfig, error) {
```
Would you mind to give an example of the config file that we are able to parse in the comment of this function?

hariohmprasath replied:
Added a sample json

hariohmprasath committed:
Incorporating code review comments

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -39,6 +43,15 @@ type BuildArguments struct {
+}
+
+const(
+	CredStoreECRLogin = "ecr-login" // set on `credStore` attribute in docker configuration file
```
```suggestion
	credStoreECRLogin = "ecr-login" // set on `credStore` attribute in docker configuration file
```
Maybe we don't need to make it public

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +172,57 @@ func imageName(uri, tag string) string {
+func parseCredFromDockerConfig(config []byte) (*dockerConfig, error) {
+	/*
+	Sample docker config file
+	"credsStore" : "ecr-login",
+	  "credHelpers": {
+	        "dummyaccountId.dkr.ecr.region.amazonaws.com": "ecr-login"
+	  },
```
```suggestion
    {
        "credsStore" : "ecr-login",
        "credHelpers": {
            "dummyaccountId.dkr.ecr.region.amazonaws.com": "ecr-login"
        }
    }
```
Could you make this json file aligned?

iamhopaul123 reviewed (approved):
Just some nits. Other than that lgtm!

hariohmprasath committed:
Update internal/pkg/exec/docker.go

Co-authored-by: Penghao He <penghaoh@amazon.com>

hariohmprasath committed:
Update internal/pkg/exec/docker.go

Co-authored-by: Penghao He <penghaoh@amazon.com>

hariohmprasath commented:
> Just some nits. Other than that lgtm!

Thanks I have accepted the suggestions, appreciate the support on this.

iamhopaul123 committed:
Merge branch 'mainline' into mainline

iamhopaul123 commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +172,59 @@ func imageName(uri, tag string) string {
+				continue
+			}
+
+			if config.CredsStore == CredStoreECRLogin || config.CredHelpers[splits[0]] == CredStoreECRLogin {
```
could you update the variable here?

hariohmprasath replied:
I just noticed the same const is used in unit test case as well. So I have updated it back to Capitol letter so it can be get used in both the places

iamhopaul123 replied:
Yeah but still they are in the same package right?

hariohmprasath replied:
You are right, been a long day I guess :) Sorry about the back and forth, its fixed now.

hariohmprasath committed:
Updating the variable because its used in test case

hariohmprasath committed:
Fixed issue around exporting credStoreLogin

hariohmprasath committed:
Fix for #2070

Added code to check for the existence of the bucket using `s3.HeadBucket` API before proceed to empty the contents

efekarakus commented on internal/pkg/aws/s3/s3.go:
```diff
@@ -105,48 +108,59 @@ func (s *S3) ZipAndUpload(bucket, key string, files ...NamedBinary) (string, err
-		}
-		_, err = s.s3Client.DeleteObjects(&s3.DeleteObjectsInput{
+
+	if isExists {
```
Can we flip the if-condition here so that the main logic is not nested?

```go
if !isExists {
    return nil
}
...
```

hariohmprasath replied:
Done

efekarakus commented on internal/pkg/aws/s3/s3.go:
```diff
@@ -161,3 +175,24 @@ func ParseURL(url string) (bucket string, key string, err error) {
+			switch aerr.Code() {
+				case notFound:
+					return false, nil
+				default:
+					return false, aerr
+			}
+		}
+		return false, err
```
nit: I believe these two statements return the same error, maybe we can simplify it like this:
```go
if aerr, ok := err.(awserr.Error); ok && aerr.Code() == notFound {
     return false, nil
}
return false, err
```

hariohmprasath replied:
Done

efekarakus commented on internal/pkg/exec/docker.go:
```diff
@@ -159,3 +172,59 @@ func imageName(uri, tag string) string {
+func (c DockerCommand) IsEcrCredentialHelperEnabled(uri string) bool {
+	// Make sure the program is able to obtain the home directory
+	splits := strings.Split(uri, "/")
+	if c.homePath != "" && len(splits) > 0 {
```
nit: can we flip this if statement to reduce nesting
```go
if c.homePath == "" || len(splits) == 0 {
     return false
}
```

hariohmprasath replied:
Done

efekarakus reviewed (approved):
Looks great thank you so much! And I'm sorry for the delay on the review 🙇 

Just small requests to flip the if-statement to reduce nesting and then the PR should auto-merge once we remove the label.

hariohmprasath committed:
Incorporating code review comments

mergify[bot] committed:
Merge branch 'mainline' into mainline
